### PR TITLE
[PuhuTv] Add new extractor

### PR DESCRIPTION
**Pull request by schrekia**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Added new extractor for free Turkish soap opera website PuhuTv.com

### CODE REVIEW

**Comment by dstftw** [comment]
Read
>[adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections

and fix all issues.

**Comment by schrekia** [comment]
It looks like failed with the DL tests, but it locally it downloads fine. Possibly because of geo-restrictions? Already applied necessary conventions. Is there any way to tell the tester to include geo-restriction-bypassing options?

**Comment by schrekia** [comment]
Looks like it fails DL tests while passing core ones. But works fine locally. Maybe geo-restrictions? Already followed conventions. Is there any way to make testers apply geo-restriction-bypassing options maybe?

File: youtube_dl/extractor/puhutv.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            if (dl.get('url').__contains__('playlist.m3u8')):
+                formats.extend(self._extract_m3u8_formats(dl.get('url'), str(dl.get('id')), 'mp4'))
+            elif (dl.get('url').__contains__('/mp4/')):
+                fmt = str(re.match(self._URL_FORMAT, dl.get('url')).group('vid_format'))
+                mp_fmts.append({
+                    'url': dl.get('url'),
+                    'format_id': 'mp4-' + fmt,
+                    'height': fmt
+                })
+        mp_fmts.sort(key=lambda x: int(x.get('height').replace('p', '')))
```
Comment (vossad01) [inline youtube_dl/extractor/puhutv.py]: `get` could return `None` resulting in,
> AttributeError: 'NoneType' object has no attribute 'replace'

Supposing that is fixed, the inner portion may still be `None`, leading to,
> TypeError: int() argument must be a string or a number, not 'NoneType'

use `int_or_none`

File: youtube_dl/extractor/puhutv.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        formats.extend(mp_fmts)
+
+        # extract video metadata
+        thumbnail = 'https://' + video_data.get('content').get('images').get('wide').get('main')
```
Comment (vossad01) [inline youtube_dl/extractor/puhutv.py]: `get` could return `None` resulting in,
> TypeError: cannot concatenate 'str' and 'NoneType' objects

File: youtube_dl/extractor/puhutv.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        description = video_data.get('description')
+        upload_date = video_data.get('created_at').split('T')[0].replace('-', '')
```
Comment (vossad01) [inline youtube_dl/extractor/puhutv.py]: `get` could return `None` resulting in,
> AttributeError: 'NoneType' object has no attribute 'split'

File: youtube_dl/extractor/puhutv.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        }
+    }
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
+        display_id = mobj.group('display_id')
+
+        video_data = self._download_json('https://puhutv.com/api/slug/' + str(display_id), display_id).get('data')
```
Comment (vossad01) [inline youtube_dl/extractor/puhutv.py]: Note `video_data` may be `None`.

File: youtube_dl/extractor/puhutv.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        video_id = str(video_data.get('id'))
+        video_title_name = video_data.get('title').get('name')
+        video_title_display_name = video_data.get('display_name')
+        video_title = video_title_name + ' ' + video_title_display_name
+
+        dl_data = self._download_json('https://puhutv.com/api/assets/' + video_id + '/videos', video_id).get('data')
```
Comment (vossad01) [inline youtube_dl/extractor/puhutv.py]: Note `dl_data` may be `None`.

File: youtube_dl/extractor/puhutv.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+        formats = []
+        mp_fmts = []
+        for dl in dl_data.get('videos'):
+            if (dl.get('url').__contains__('playlist.m3u8')):
```
Comment (vossad01) [inline youtube_dl/extractor/puhutv.py]: [You should not be calling `__contains__` directly.](http://stackoverflow.com/a/1964949/1072626)

File: youtube_dl/extractor/puhutv.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+                formats.extend(self._extract_m3u8_formats(dl.get('url'), str(dl.get('id')), 'mp4'))
+            elif (dl.get('url').__contains__('/mp4/')):
```
Comment (vossad01) [inline youtube_dl/extractor/puhutv.py]: [You should not be calling `__contains__` directly.](http://stackoverflow.com/a/1964949/1072626)

File: youtube_dl/extractor/puhutv.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+                fmt = str(re.match(self._URL_FORMAT, dl.get('url')).group('vid_format'))
```
Comment (vossad01) [inline youtube_dl/extractor/puhutv.py]: `get` could return `None` resulting in,
> AttributeError: 'NoneType' object has no attribute 'group'

File: youtube_dl/extractor/puhutv.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+            'id': 'isler-gucler-1-bolum-izle',
+            'title': 'İşler Güçler 1. Sezon 1. Bölüm',
+            'ext': 'mp4',
+            'thumbnail': r're:^https?://.*\.(JPG|jpg)$',
+            'upload_date': '20160719',
+            'description': 'Ahmet, başrol oynadığı günlerin özlemini çekmekte, mazideki başarılarının göz ardı edilmesinden şikayet etmektedir. Murat, mesleğinin gereklerini yapmaya çalışırken, kendisini terketmiş olan eski sevgilisiyle beraber çalışmanın zorluklarını yaşamaktadır. Sadi, kapı gibi gümrük memurluğunu oyuncu olmak için bırakmış, para dertleri ve ailevi sorunlarla uğraşmaktadır. Figürasyon yaptığı bir tarihi diziden başka gelir kaynağı olmayan Sadi için \'Meslek Hikayeleri\'nin yayınlanması ve ücretlerini almaları çok önemlidir. \nBu sıra da programın Yapımcısı ve yapım ekibi işleri kolaylaştırmamakta güçleştirmektedir. İlginç aile üyeleri, değişik menajerleri ve olmazsa olmaz komşuları ile üçlümüzün hayatları bütün tuhaflıkları ile devam etmektedir. \'Meslek Hikayeleri\' için çektikleri ve yayınlanmayan bölümler içlerine dert olan üçlü, yayınlanabilecek kalitede bir bölüm için kolları sıvarlar. Ve ellerinden geleni de yaparlar.'
```
Comment (dstftw) [inline youtube_dl/extractor/puhutv.py]: `'description': 'md5:...'`.

File: youtube_dl/extractor/puhutv.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+import re
+
+from .common import InfoExtractor
+
+
+class PuhuTvIE(InfoExtractor):
+    _VALID_URL = r'https?:\/\/(?:www\.)?puhutv\.com\/(?P<display_id>[^/]+-izle)(\?[^/?]+)?'
```
Comment (dstftw) [inline youtube_dl/extractor/puhutv.py]: No escape for `\` needed. `(\?[^/?]+)?` makes no sense at the end.

File: youtube_dl/extractor/puhutv.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+            'upload_date': '20160719',
+            'description': 'Ahmet, başrol oynadığı günlerin özlemini çekmekte, mazideki başarılarının göz ardı edilmesinden şikayet etmektedir. Murat, mesleğinin gereklerini yapmaya çalışırken, kendisini terketmiş olan eski sevgilisiyle beraber çalışmanın zorluklarını yaşamaktadır. Sadi, kapı gibi gümrük memurluğunu oyuncu olmak için bırakmış, para dertleri ve ailevi sorunlarla uğraşmaktadır. Figürasyon yaptığı bir tarihi diziden başka gelir kaynağı olmayan Sadi için \'Meslek Hikayeleri\'nin yayınlanması ve ücretlerini almaları çok önemlidir. \nBu sıra da programın Yapımcısı ve yapım ekibi işleri kolaylaştırmamakta güçleştirmektedir. İlginç aile üyeleri, değişik menajerleri ve olmazsa olmaz komşuları ile üçlümüzün hayatları bütün tuhaflıkları ile devam etmektedir. \'Meslek Hikayeleri\' için çektikleri ve yayınlanmayan bölümler içlerine dert olan üçlü, yayınlanabilecek kalitede bir bölüm için kolları sıvarlar. Ve ellerinden geleni de yaparlar.'
+        }
+    }
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
+        display_id = mobj.group('display_id')
```
Comment (dstftw) [inline youtube_dl/extractor/puhutv.py]: `self._match_id`.

File: youtube_dl/extractor/puhutv.py (hunk 4 above)
Comment (dstftw) [inline youtube_dl/extractor/puhutv.py]: `display_id` is already string.

**Comment by dstftw** [comment]
Superseded by #16269.

**Pull request closed without merging**